_posts/2021-10-01-debezium-server-iceberg.adoc

```diff
@@ -0,0 +1,163 @@
+
+== Debezium Server Iceberg
+
+**Debezium Server Iceberg** project implements new consumer which uses Iceberg Java API to consume events.
+Iceberg consumer converts received events to iceberg format and commits them to destination iceberg table. With the consumer it's possible to use any cloud store/any catalog supported by iceberg.
+
+on high level data processing works like below ::
+for each batch of events
+* it groups events per event destination,
+* for each destination, events are converted to iceberg records. at this step event schema used to map data to iceberg record (`debezium.format.value.schemas.enable` should be enabled(true)).
+* After events converted to iceberg records, they are saved to iceberg parquet files(iceberg data and delete files(for upsert)).
+* as a last step these files are committed to destination table(uploaded to destination storage)) using iceberg java API.
+
+If destination table not found in the destination catalog consumer will try to create it using event schema and key schema(record key/PK).
+
+image::../assets/images/debezium-iceberg.png[Debezium-Iceberg]
+
+Currently, Iceberg Consumer supports only json events. With json events event schema must be enabled to do correct type conversion to iceberg record.
+
+example configuration::
+[source,properties]
+----
+debezium.sink.type=iceberg
+# run with append mode
+debezium.sink.iceberg.upsert=false
+debezium.sink.iceberg.upsert-keep-deletes=true
+# iceberg
+debezium.sink.iceberg.table-prefix=debeziumcdc_
+debezium.sink.iceberg.table-namespace=debeziumevents
+debezium.sink.iceberg.fs.defaultFS=s3a://S3_BUCKET);
+debezium.sink.iceberg.warehouse=s3a://S3_BUCKET/iceberg_warehouse
+debezium.sink.iceberg.type=hadoop
+debezium.sink.iceberg.catalog-name=mycatalog
+debezium.sink.iceberg.catalog-impl=org.apache.iceberg.hadoop.HadoopCatalog
+# enable event schemas
+debezium.format.value.schemas.enable=true
+debezium.format.value=json
+# complex nested data types are not supported, do event flattening. unwrap message!
+debezium.transforms=unwrap
+debezium.transforms.unwrap.type=io.debezium.transforms.ExtractNewRecordState
+debezium.transforms.unwrap.add.fields=op,table,source.ts_ms,db
+debezium.transforms.unwrap.delete.handling.mode=rewrite
+debezium.transforms.unwrap.drop.tombstones=true
+----
+
+=== Update, Append modes
+
+By default, Iceberg sink is running with upsert mode `debezium.sink.iceberg.upsert=true`. When a row updated on source table destination row replaced with the new updated version, and deleted records are deleted from destination. With upsert mode data at destination kept identical to source data. Update mode uses iceberg equality delete feature and creates delete files using key of the debezium event(PK). To avoid duplicate data deduplication is done on each batch and only last version of the record kept.
+
+Note: For the tables without record key(PK) operation mode falls back to `append` even configuration is set to upsert mode.
```

Comment:
Instead of this pathological case, can you perhaps describe how to use append mode in general? I think that's interesting for many users who would like to keep the entire change history for a record in Iceberg.